File: python/core/sampling.py
from typing import Any

from anthropic import AsyncAnthropic
from mcp import types
from mcp.shared.context import RequestContext

from core.claude import Claude


def make_sampling_callback(claude_service: Claude):
    """Builds a sampling callback that fulfills the server's `create_message`
    requests by calling Claude through the Anthropic SDK.

    Uses its own `AsyncAnthropic` client rather than the shared sync `Claude`
    service so the call doesn't block the event loop while other async work
    (other MCP traffic, the CLI prompt) is in flight.
    """
    anthropic_client = AsyncAnthropic()
    model = claude_service.model

    async def sampling_callback(
        context: RequestContext[Any, Any],
        params: types.CreateMessageRequestParams,
    ) -> types.CreateMessageResult | types.ErrorData:
        try:
            messages = [
                {
                    "role": message.role,
                    "content": message.content.text
                    if isinstance(message.content, types.TextContent)
                    else str(message.content),
                }
                for message in params.messages
            ]

            create_kwargs: dict[str, Any] = {
                "model": model,
                "messages": messages,
                "max_tokens": params.maxTokens,
            }
            if params.systemPrompt:
                create_kwargs["system"] = params.systemPrompt
            if params.temperature is not None:
                create_kwargs["temperature"] = params.temperature
            if params.stopSequences:
                create_kwargs["stop_sequences"] = params.stopSequences

            response = await anthropic_client.messages.create(**create_kwargs)

            text = "\n".join(
                block.text for block in response.content if block.type == "text"
            )

            return types.CreateMessageResult(
                role="assistant",
                model=model,
                content=types.TextContent(type="text", text=text),
            )
        except Exception as e:
            return types.ErrorData(code=types.INTERNAL_ERROR, message=str(e))

    return sampling_callback

File: typescript/src/core/sampling.ts
import Anthropic from "@anthropic-ai/sdk";
import type { MessageParam } from "@anthropic-ai/sdk/resources/messages";
import {
  ErrorCode,
  McpError,
  type CreateMessageRequest,
  type CreateMessageResult,
  type SamplingMessage,
} from "@modelcontextprotocol/sdk/types.js";
import { Claude } from "./claude.js";

// `SamplingMessage.content` can be a single content block or an array of them.
function contentToText(content: SamplingMessage["content"]): string {
  const blocks = Array.isArray(content) ? content : [content];
  return blocks
    .filter((block) => block.type === "text")
    .map((block) => (block as { text: string }).text)
    .join("\n");
}

// Builds a handler for the server's `sampling/createMessage` requests,
// fulfilling them by calling Claude through the Anthropic SDK. Equivalent to
// `core/sampling.py`. The Anthropic TS SDK is async/non-blocking by default,
// so (unlike the Python port) there's no separate sync/async client concern
// here — a plain `Anthropic()` client is enough.
export function makeSamplingHandler(claudeService: Claude) {
  const anthropicClient = new Anthropic();
  const model = claudeService.getModel();

  return async (request: CreateMessageRequest): Promise<CreateMessageResult> => {
    try {
      const { params } = request;

      const messages: MessageParam[] = params.messages.map(
        (message: SamplingMessage) => ({
          role: message.role,
          content: contentToText(message.content),
        })
      );

      const response = await anthropicClient.messages.create({
        model,
        messages,
        max_tokens: params.maxTokens,
        ...(params.systemPrompt ? { system: params.systemPrompt } : {}),
        ...(params.temperature !== undefined
          ? { temperature: params.temperature }
          : {}),
        ...(params.stopSequences && params.stopSequences.length > 0
          ? { stop_sequences: params.stopSequences }
          : {}),
      });

      const text = response.content
        .filter((block) => block.type === "text")
        .map((block) => (block as { text: string }).text)
        .join("\n");

      return {
        role: "assistant",
        model,
        content: { type: "text", text },
      };
    } catch (error) {
      throw new McpError(
        ErrorCode.InternalError,
        error instanceof Error ? error.message : String(error)
      );
    }
  };
}
